### PR TITLE
F# Unit tests

### PR DESCRIPTION
**Pull request by mereta**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,4 +8,11 @@ riderModule.iml
 metrics.json
 plot.html
 results.bin
-.vscode/
\ No newline at end of file
+.vscode/
+/.vs/road-to-orleans/v16
```
Comment (mrblonde91) [inline .gitignore:12]: Think `**/.vs` will ignore all the vs folders in repo.

Comment (mereta) [reply .gitignore]: Good point.

**Comment by mrblonde91** [comment]
I approve anyway :D

File: 5/Library/Library.fs [hunk 2]
```diff
@@ -2,4 +2,11 @@
 
 module Say =
     let hello name =
-        sprintf "Hello, \"%s\"! Your name is %i characters long." name (String.length name)
+        let _input =
```
Comment (PiotrJustyna) [inline 5/Library/Library.fs:5]: Nitpicking, but any reason for this underscore?

File: 5/UnitTests/src/UnitTests/Tests.fs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+open Client
+
+[<Fact>]
+let ``Library Valid String Input Test`` () =
+    let result = Say.hello("Mereta")
+    Assert.Equal("Hello, \"Mereta\"! Your name is 6 characters long.", result)
+
+[<Fact>]
+let ``Library Empty String Input Throws Exception`` () =
+    let methodCall = (fun () -> Say.hello "" |> ignore)
```
Comment (PiotrJustyna) [inline 5/UnitTests/src/UnitTests/Tests.fs:17]: Technically it's not a method call, but a function (uncalled yet), so maybe it would make more sense to either extract it as a reusable function or as an inline lambda used in lines e.g. 18, 24, etc.

File: 5/UnitTests/src/UnitTests/Tests.fs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    Assert.Throws<System.ArgumentException>(methodCall)
+
+
+[<Fact>]
+let ``Library Exception Messgage Test`` () = 
+    let methodCall = (fun () -> Say.hello "" |> ignore)
+    let ex = Assert.Throws<System.ArgumentException>(methodCall);
+    Assert.Equal("string cannot be null or whitespace (Parameter 'name')" , ex.Message);
```
Comment (PiotrJustyna) [inline 5/UnitTests/src/UnitTests/Tests.fs:25]: This and the previous test can be probably rolled into one.

File: 5/UnitTests/src/UnitTests/Tests.fs [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+[<Fact>]
+let ``Grains Service Test Returns Correct Output`` () =
+    let _grainService = {
+        new IHelloWorld with 
+            member _this.SayHello(input) = 
+                match input with 
+                    "Mereta" -> Task.FromResult("Hello, \"Mereta\"! Your name is 6 characters long.") }
+    let input, output = "Mereta",  Task.FromResult("Hello, \"Mereta\"! Your name is 6 characters long.")
+    let result = _grainService.SayHello(input)
+    Assert.Equal(output.Result, result.Result)
```
Comment (PiotrJustyna) [inline 5/UnitTests/src/UnitTests/Tests.fs:37]: We probably don't need this test as all it does is it verifies the mock.

File: 5/UnitTests/src/UnitTests/Tests.fs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+
+[<Fact>]
+let ``Hello World Client Returns Task`` () =
+    let _clientService = Mock<HelloWorldClientHostedService>().Setup(fun me -> <@ me.StartAsync(any()) @> ).Returns(Task.CompletedTask).Create()
+    let output = _clientService.StartAsync(any())
+    Assert.Equal(typeof<Task>, output.GetType().BaseType)
```
Comment (PiotrJustyna) [inline 5/UnitTests/src/UnitTests/Tests.fs:44]: We don't need it either, there's no need for mocks in our scenario. Ideally, I can only imagine 3 tests we could use:

1. correct, 1+ characters long input
2. whitespace input "\t\n.  \r"
3. empty input

Feel free to add more tests, but that the minimum we would need.

File: 5/UnitTests/src/UnitTests/Tests.fs [hunk 7]
```diff
@@ -0,0 +1,27 @@
+module UnitTests.Tests
+
+open Xunit
+open Library
+let func y = fun () -> Say.hello y
```
Comment (PiotrJustyna) [inline 5/UnitTests/src/UnitTests/Tests.fs:5]: Is this not used?

**Review by PiotrJustyna — APPROVED** [review]
Very nice, thanks. Just one small comment.
